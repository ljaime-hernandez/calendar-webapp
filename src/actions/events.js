import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { fixEvents } from "../helpers/fixEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {

    return async(dispatch, getState) => {
        
        // we use the getState to retrieve the uid and name in the store dispatched to our authReducer when
        // the user logged on its account
        const { uid, name } = getState().auth;

        try {

        // the fetchWithToken function is used as, for this request to work, an authorized user must be logged in, 
        // if the user is authenticated and the token is still valid, then the event data will go through some checks
        // on the backend middlewares, to finally be registered in our mongo database. The first argument will be the 
        // endpoint which is 'events' and the third argument must be a POST
            const res = await fetchWithToken('events', event, 'POST');
            const body = await res.json();

                // if the response is correct, then we will add the uid and name of the user for it to be used as reference
                // on the eventModal, which will help us register the event using this user as reference for whenever we want
                // to retrieve all events registered on it
                if(body.ok){
                    event.id = body.event.id;
                    event.user = {
                        _id: uid,
                        name: name
                    };
                    // the event information will be dispatched to an array of events which we will be able to track from the store
                    // if necessary
                    dispatch(eventAddNew(event));
                };
            
        } catch (error) {
            console.log(error);
        };
    };
};

export const eventStartLoading = () => {

    return async(dispatch) => {

        try {
            
            // The fetchWithToken will do a GET request by default and will use the endpoint to receive all the events
            // received from the calendar backend for them to later be displayed in the users calendar display.
            const resp = await fetchWithToken('events');
            const body = await resp.json();
            // The start and end dates on the events might not work properly, so we need a helper for us to change the
            // dates registered in the mongo database and modify them with moment.js for us to use them properly
            const events = fixEvents(body.events);
            
            dispatch(eventLoaded(events));

        } catch (error) {
            console.log(error);
        };
    };
} ;

export const eventStartUpdate = (event) => {

    return async(dispatch) => {

        try {
            
            // this fetch will use the PUT as method to update the event by using its id and the events endpoint as
            // reference, if the event is found then the calendar backend will use a mongo method for the event to be
            // updated
            const resp = await fetchWithToken(`events/${event.id}`, event, 'PUT');
            const body = await resp.json();

            if(body.ok) {
                dispatch(eventUpdated(event));
                dispatch(eventStartLoading());
            } else {
                // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
                // which will be attached to the body response 
                Swal.fire('Error', body.msg, 'error');
            };

        } catch (error) {
            console.log(error);
        };
    };
};

export const eventStartDelete = () => {

    return async(dispatch, getState) => {

        // we use the getState to retrieve the id in the store dispatched to our calendarReducer when
        // the user clicks on an specific event
        const { id } = getState().calendar.activeEvent; 

        try {
            
            // Similar to the eventStartUpdate, this fetch will use the DELETE as method to delete the event by using 
            // its id and the events endpoint as reference, if the event is found then the calendar backend will use 
            // a mongo method for the event to be deleted and no additional event data needs to be sent as the deletion
            // only requires the id for it
            const resp = await fetchWithToken(`events/${id}`, {}, 'DELETE');
            const body = await resp.json();

            if(body.ok) {
                dispatch(eventDeleted());
                dispatch(eventStartLoading());
                dispatch(eventClearActive());
            } else {
                // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
                // which will be attached to the body response 
                Swal.fire('Error', body.msg, 'error');
            };
        } catch (error) {
            console.log(error);
        };
    };
};

export const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () => ({
    type: types.eventClearActive
});

const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
})

const eventDeleted = () => ({
    type: types.eventDeleted
})

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
})

export const eventLogout = () => ({
    type: types.eventLogout
}) 