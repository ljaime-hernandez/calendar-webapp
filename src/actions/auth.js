import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
    
    return async(dispatch) => {

        // the fetchWithoutToken function is used as the user is not logged in and we are about to create a token
        // for it if the user exists in our database, the first argument will be the endpoint which is the
        // auth and the third argument must be a POST
        const resp = await fetchWithoutToken('auth', {email, password}, 'POST');
        const body = await resp.json();

        // if we receive the correct response, we will use the localStorage to temporarily place our received token,
        // by default the token will last for 2 hours, but it can be renewed with some of the functions created on our
        // calendar, we also place the specific time where the token was placed in the token-init-date portion of the
        // localStorage to know if the token is still valid.
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            // the dispatch will be sent to the store for it to be registered on the reducer
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));

        } else {
            // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
            // which will be attached to the body response 
            Swal.fire('error', body.msg, 'error');
        };
    };
};

// The startRegister function will not only register a new user on the database but it will also 
// automatically log the user in.
export const startRegister = (name, email, password) => {

    return async(dispatch) => {

        // The fetchWithoutToken function is used as the user is not logged in and we are about to create a token
        // for it. The first argument will be the endpoint which is the auth/new and the third argument must be a POST
        const resp = await fetchWithoutToken('auth/new', {name, email, password}, 'POST');
        const body = await resp.json();

        // If we receive the correct response, we will use the localStorage to temporarily place our received token,
        // by default the token will last for 2 hours, but it can be renewed with some of the functions created on our
        // calendar, we also place the specific time where the token was placed in the token-init-date portion of the
        // localStorage to know if the token is still valid.
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(register({
                uid: body.uid,
                name: body.name
            }));

            // the dispatch will be sent to the store for it to be registered on the reducer
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));

        } else {
            // the error message displayed in the Sweetalert popup is retrieved from the calendar backend, 
            // which will be attached to the body response 
            Swal.fire('Error', body.msg, 'error');
        };
    };
};

export const startChecking = () => {

    return async(dispatch) => {

        // the endpoint of the fethWithToken uses the renew portion of it for the user to retrieve an updated token for the 
        // user to continue its operations on the webapp for a longer period of time
        const resp = await fetchWithToken('auth/renew');
        const body = await resp.json();

        // If we receive the correct response, we will use the localStorage to temporarily place our received token,
        // by default the token will last for 2 hours, but it can be renewed with some of the functions created on our
        // calendar, we also place the specific time where the token was placed in the token-init-date portion of the
        // localStorage to know if the token is still valid.
        if(body.ok){
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(checkDone());
            // the dispatch will be sent to the store for it to be registered on the reducer
            dispatch(login({
                uid: body.uid,
                name: body.name
            }));
        } else {
            dispatch(checkDone());
        };
    };
};

export const startLogout = () => {

    return (dispatch) => {

        // once the user logs out, we will clean every piece of information from the localStorage
        localStorage.clear();
        dispatch(eventLogout());
        dispatch(logout());
    }
}

const login = (user) => ({
    type: types.authLogin,
    payload: user
})

const register = (user) => ({
    type: types.authStartRegister,
    payload: user
})

const checkDone = () => ({type: types.authCheckDone});

const logout = () => ({type: types.authLogout});