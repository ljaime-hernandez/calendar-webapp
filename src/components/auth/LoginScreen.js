import React from 'react';
import Swal from 'sweetalert2';
import { useDispatch } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { startLogin, startRegister } from '../../actions/auth';

export const LoginScreen = () => {

    const dispatch = useDispatch();
    
    // the values on this form are set by default so the user can check the functionality of the webapp without creating
    // their own user, but it can receive different information if requested
    const [values, handleInputChange, reset] = useForm({
        loginEmail: 'email@email.com',
        loginPassword: '122345'
    });

    // the values on this form are set empty so the user can create its account
    const [registerValues, handleRegisterInputChange, registerReset] = useForm({
        registerName: '',
        registerEmail: '',
        registerPassword: '',
        registerConfirmPassword: ''
    });

    // the values of both the login and register forms are retrieved for use to use them as reference for any operation
    // we require in the future 
    const { loginEmail, loginPassword } = values;
    const { registerName, registerEmail, registerPassword, registerConfirmPassword } = registerValues;

    const handleLogin = (e) => {
        e.preventDefault();
        // dispatches the function using the first form to log the user in
        dispatch(startLogin( loginEmail, loginPassword ));
        reset();
    }

    const handleRegister = (e) => {
        e.preventDefault();

        // compares the registration password as they have to be equal, if not then the user wont be able
        // to submit the form, the calendar backend also has a secondary filter which will force the user to
        // submit a password of at least 6 characters, displayed in a sweetalert popup with the respective message
        if(registerPassword !== registerConfirmPassword){
            return Swal.fire('error', 'Password does not match with password confirmation', 'error');
        }

        // will register the users information if properly submitted onto out mongo database. If successful,
        // the user will also be logged in to the webpage
        dispatch(startRegister(registerName, registerEmail, registerPassword ));
        // Resets the values of the register form
        registerReset();
    }

    return (
        <div className='container mt-5 pt-5'>
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Access</h3>
                    <form onSubmit={handleLogin}>
                        <div className="form-group">
                            <input 
                                type="text"
                                className="form-control"
                                placeholder="Email"
                                name="loginEmail"
                                value={loginEmail}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password"
                                name="loginPassword"
                                value={loginPassword}
                                onChange={handleInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input 
                                type="submit"
                                className="btnSubmit"
                                value="Login" 
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Register</h3>
                    <form onSubmit={handleRegister}>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Name"
                                name="registerName"
                                value={registerName}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                                name="registerEmail"
                                value={registerEmail}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password should be at least 6 characters long"
                                name="registerPassword"
                                value={registerPassword}
                                onChange={handleRegisterInputChange}
                            />
                        </div>

                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Confirm your password"
                                name="registerConfirmPassword"
                                value={registerConfirmPassword}
                                onChange={handleRegisterInputChange}
                            />
                        </div>

                        <div className="form-group">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Create Account" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}