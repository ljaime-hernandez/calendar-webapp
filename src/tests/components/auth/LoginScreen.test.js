import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import '@testing-library/jest-dom';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin } from '../../../actions/auth';


jest.mock('../../../actions/auth', () => ({
    startLogin: jest.fn()
}))

const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider
        store={store}
    >
        <LoginScreen/>
    </Provider>
)

describe('Tests on LoginScreen component', () => {

    test('should render properly', () => { 

        expect(wrapper).toMatchSnapshot();
    });

    test('should call login dispatch', () => { 

        wrapper.find('input[name="loginEmail"]').simulate('change', {
            target: {
                name: 'loginEmail',
                value: 'email@emaill.com'
            }
        });

        wrapper.find('input[name="loginPassword"]').simulate('change', {
            target: {
                name: 'loginPassword',
                value: '122345'
            }
        });

        wrapper.find('form').at(0).prop('onSubmit')({
            preventDefault(){}
        });

        expect(startLogin).toHaveBeenCalledWith('email@emaill.com', '122345');
    });

    test('should not register if password confirm does not match the password', () => {

        wrapper.find('input[name="registerPassword"]').simulate('change', {
            target: {
                name: 'registerPassword',
                value: 'email@emaill.com'
            }
        });

        wrapper.find('input[name="registerConfirmPassword"]').simulate('change', {
            target: {
                name: 'registerConfirmPassword',
                value: '122345'
            }
        });
    })

})