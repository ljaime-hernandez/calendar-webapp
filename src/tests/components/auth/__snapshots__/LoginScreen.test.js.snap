// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on LoginScreen component should render properly 1`] = `
<div
  className="container mt-5 pt-5"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Access
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value="email@email.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value="122345"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value="Miguel"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value="email@email.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value="122345"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerConfirmPassword"
            onChange={[Function]}
            placeholder="Confirm your password"
            type="password"
            value="122345"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Create Account"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
